### PR TITLE
Upgrade to Clang 17.0.6 in the C++20 buildbot

### PR DESCRIPTION
**Pull request by ilya-biryukov**
See https://github.com/llvm/llvm-project/pull/72348 for some context on why we need this upgrade.

### CODE REVIEW

**Comment by ilya-biryukov** [comment]
I have checked that this works, but I believe we should wait until the [RFC](https://discourse.llvm.org/t/rfc-clang-17-0-6-would-be-minimum-version-to-build-llvm-in-c-20/75345/12) gets discussed properly before landing this.

Putting Aaron as the reviewer for this change to stamp when the discussion has reached some conclusion.

**Review by AaronBallman — APPROVED** [review]
LGTM assuming there are no surprises on the RFC in the next few weeks.

**Review by usx95 — APPROVED** [review]
LGTM. [As Aaron mentioned on the RFC](https://discourse.llvm.org/t/rfc-clang-17-0-6-would-be-minimum-version-to-build-llvm-in-c-20/75345/14), we will wait for two more weeks for comments on the RFC.
If no concerns are raised, we can merge this on 21-Dec-2023.

**Pull request merged**
